src/v/storage/log_manager.cc

```diff
@@ -94,14 +95,55 @@ ss::future<> log_manager::stop() {
       .then([this] { return _batch_cache.stop(); });
 }
 
-static inline logs_type::iterator find_next_non_compacted_log(logs_type& logs) {
+inline logs_type& log_manager::modify_logs() {
+    _logs_version++;
+    return _logs;
+}
+
+static inline logs_type::iterator find_next_non_compacted_log(
+  logs_type& logs, typename logs_type::iterator start_iter) {
     using bflags = log_housekeeping_meta::bitflags;
     return std::find_if(
-      logs.begin(), logs.end(), [](const logs_type::value_type& l) {
+      start_iter, logs.end(), [](const logs_type::value_type& l) {
           return bflags::none == (l.second.flags & bflags::compacted);
       });
 }
 
+/**
+ * `housekeeping_scan` attempts to scan over `_logs` in a single pass.
+ * This is achieved by using `_logs_version` a counter that is incremented
+ * everytime `_logs` is modified. If the counter has not been modified after
+ * seastar resumes the scan then we know our iterator is still valid and can
+ * continue scanning over `_logs` with it. Otherwise if it has been modified
+ * our iterator is now invalid and we have to re-scan over `_logs` to establish
+ * a new valid iterator.
+ */
+ss::future<>
+log_manager::housekeeping_scan(model::timestamp collection_threshold) {
+    using bflags = log_housekeeping_meta::bitflags;
+
+    auto current_iter = find_next_non_compacted_log(_logs, _logs.begin());
+    auto current_version = _logs_version;
+
+    while (current_iter != _logs.end()) {
+        current_iter->second.flags |= bflags::compacted;
+        current_iter->second.last_compaction = ss::lowres_clock::now();
+        co_await current_iter->second.handle.compact(compaction_config(
+          collection_threshold,
+          _config.retention_bytes(),
+          _config.compaction_priority,
+          _abort_source));
+
+        if (current_version != _logs_version) {
```

Comment:
props on detecting changes here. we do something similar in the memory claimer and generally it works really well.

the downside to this is that it's prone to developer mistakes (e.g. not going through the modify() wrapper). but unlike the memory reclaimer, this loop doesn't seem to be in the hot path, it merely needs to be efficient for large collections. so i'm wondering if there are other approaches, here are a couple examples:

1. what if we kept the `_logs` collection the way it is now as the source of true, but constructed a secondary index as a separate collection of logs that marked as needing to be compacted. in the simple case we'd initialize both containers to be the same, and pop an element out of secondary index to compact it. when the container is empty we'd scan the source of truth to repopulate.

2. another option would be to store the logs in a boost instructive list ordered by least-recently-compacted. the O(1) index from ntp->log would still be used, but each time we compacted we'd stick the compated log at the head of the intrusive list (very fast).

I'm not opposed to this particular trick with the version, but I'm just curious if there are other points in the design space worth thinking about.

Reply:
We can avoid developer mistakes by doing something like;
```cpp
class logs_protected {
     logs_type _logs;
public:
    logs_type& modify_logs();
    logs_type const& read_logs();
};
```

It the space complexity isn't an issue keeping a separate queue of logs from the `_logs` hash map sounds like a good idea. We could add the log to the queue at the same time we add it to `_logs`. Then we could remove a set number per housekeeping and then re-add them to the back of the queue after compaction.

Reply:
>     logs_type const& read_logs();

i can still obtain an iterator from this interface, hold on to it, and get myself it trouble. but it's better than nothing!

> However, removing logs would still invalidate iterators. 

i haven't thought about it in detail. but i think the key is to not hold on to any iterators. either by looking up by ntp in both containers, or representing the relationships explicitly.

One option is something like this:

```c++
struct log_wrapper {
   intrusive_list_link;
   log l;
};

map<ntp, std::unique_ptr<log_wrapper>> logs;
intrusive_list<log_wrapper>> least_recently_compacted;
```

* to find a a log by ntp, use logs[ntp]
* to find the next candidate to compact, look at the first element on the lrc list
* when a compaction target is chosen, move it to the end of the list
* when a log is deleted, delete it from both lists

the key thing i'm assuming here is that the least-recently-compacted organization provides the search ordering that you want. if not, i may be leading you down a bad path here.

do you or @jcsp have a target number of elements that will be in the container? the thing is that absl::flat_hash_map may cause problems for large numbers of partitions. but we can swap in absl::btree_map in this case. intrusive list doesn't have any sizing problems related to memory allocations.

Reply:
Went ahead and implemented the intrusive list version of this. As you've said it definitely reduces potential developer errors in the future. I think we have some goals for total size related to how many partitions we can run on a single node. But that size is nowhere's near `absl::flat_hash_map` being an issue.

Reply:
Unfortunately intrusive lists don't work well when storing their nodes in a container that dynamically reallocates it's self. I've had to switch to a `std::list` unfortunately. This means there is a lot more memory fragmentation though unfortunately along with additional logic to keep `_logs` and `_logs_list` in sync with one another. Let me know what you think.